model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.connections.jpa.updater.liquibase.custom;
+
+import liquibase.exception.CustomChangeException;
+import liquibase.statement.core.InsertStatement;
+import liquibase.structure.core.Table;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.keycloak.protocol.saml.util.ArtifactBindingUtils.computeArtifactBindingIdentifierString;
+
+public class JpaUpdate14_0_0_MigrateSamlArtifactAttribute extends CustomKeycloakTask {
+
+    private static final String SAML_ARTIFACT_BINDING_IDENTIFIER = "saml.artifact.binding.identifier";
+
+    private final Map<String, String> clientIds = new HashMap<>();
+
+    @Override
+    protected void generateStatementsImpl() throws CustomChangeException {
+        extractClientsData("SELECT ID, CLIENT_ID FROM " + getTableName("CLIENT") + " " +
+                "WHERE PROTOCOL='saml'");
```

Comment:
```suggestion
        extractClientsData("SELECT c.ID, c.CLIENT_ID FROM " + getTableName("CLIENT") + " c " +
                " LEFT JOIN " + getTableName("CLIENT_ATTRIBUTES") + " ca ON c.CLIENT_ID = ca.CLIENT_ID AND ca.name = '" + SAML_ARTIFACT_BINDING_IDENTIFIER + "' " +
                "WHERE c.PROTOCOL='saml' AND ca.name IS NULL");
```

Reply:
Thanks, that is nice improvement. I will add it.